core/drivers/crypto/marvell/authenc.c

```diff
@@ -0,0 +1,490 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2025, Marvell
+ */
+
+#include <assert.h>
+#include <crypto/crypto.h>
+#include <crypto/crypto_impl.h>
+#include <crypto/internal_aes-gcm.h>
+#include <drvcrypt.h>
+#include <drvcrypt_authenc.h>
+#include <initcall.h>
+#include <stdlib.h>
+#include <string.h>
+#include <string_ext.h>
+#include <tee_api_types.h>
+#include <tee/cache.h>
+#include <utee_defines.h>
+#include <util.h>
+
+#include "mrvl_ehsm_cryp.h"
+
+struct mrvl_ae_ctx {
+	struct crypto_authenc_ctx a_ctx;
+	struct mrvl_cryp_context cryp_ctx;
+};
+
+static struct mrvl_ae_ctx *to_mrvl_ae_ctx(struct crypto_authenc_ctx *ctx)
+{
+	assert(ctx);
+
+	return container_of(ctx, struct mrvl_ae_ctx, a_ctx);
+}
+
+static TEE_Result mrvl_ae_initialize(struct drvcrypt_authenc_init *dinit)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	struct mrvl_ae_ctx *c = to_mrvl_ae_ctx(dinit->ctx);
+	struct mrvl_cryp_context *ctx = &c->cryp_ctx;
+
+	mrvl_ehsm_cryp_lock();
+
+	ctx->iv_len = dinit->nonce.length;
+	ctx->iv_data = mem_alloc(ctx->iv_len);
+	if (!ctx->iv_data) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+	memcpy(ctx->iv_data, dinit->nonce.data, ctx->iv_len);
+
+	ctx->aad_len = dinit->aad_len;
+	ctx->aad_cur_idx = 0;
+	ctx->aad_data = mem_alloc(ctx->aad_len);
+	if (!ctx->aad_data) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto out_aad_err;
+	}
+
+	ctx->op_enc_dec = dinit->encrypt;
+	ctx->tag_len = dinit->tag_len;
+
+	ctx->key_len = dinit->key.length;
+	ctx->key = mem_alloc(ctx->key_len);
+	if (!ctx->key) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto out_key_err;
+	}
+	memcpy(ctx->key, dinit->key.data, ctx->key_len);
+
+	ret = mrvl_ehsm_aes_gcm_init(ctx->op_enc_dec, (void *)ctx->key,
+				     ctx->key_len, ctx->aad_len,
+				     ctx->tag_len, ctx->iv_len,
+				     ctx->iv_data, 0);
+	if (ret != TEE_SUCCESS)
+		goto out_init_err;
+
+	ret = mrvl_ehsm_aes_context_store(&ctx->context_id);
+	if (ret != TEE_SUCCESS)
+		goto out_init_err;
+
+	ctx->is_new = true;
+
+	goto out;
+
+out_init_err:
+	free(ctx->key);
+out_key_err:
+	free(ctx->aad_data);
+out_aad_err:
+	free(ctx->iv_data);
```

Comment:
Prefer use a branch label that tells what is done on the error path rather than why we've branched here:
```c
err_free_key:
	free(ctx->key);
err_free_aad:
	free(ctx->aad_data);
err_free_iv:
	free(ctx->iv_data);
```
Ditto for `mrvl_ae_update_payload(), `mrvl_ae_enc_final()` and `mrvl_ae_dec_final()`.